File: rust/psibase/src/services/auth_delegate.rs
#[crate::service(name = "auth-delegate", dispatch = false, psibase_mod = "crate")]
#[allow(non_snake_case, unused_variables)]
mod service {
    use crate::{services::transact::ServiceMethod, AccountNumber, Claim};

    /// This is an implementation of the standard auth service interface defined in [SystemService::AuthInterface]
    ///
    /// This action is automatically called by `transact` when an account using this auth service submits a
    /// transaction.
    ///
    /// This action forwards verification to the owning account
    #[action]
    fn checkAuthSys(
        flags: u32,
        requester: AccountNumber,
        sender: AccountNumber,
        action: ServiceMethod,
        allowedActions: Vec<ServiceMethod>,
        claims: Vec<Claim>,
    ) -> bool {
        unimplemented!()
    }

    /// Gets the owner account of the specified account
    #[action]
    fn getOwner(account: AccountNumber) -> AccountNumber {
        unimplemented!()
    }

    /// This is an implementation of the standard auth service interface defined in [SystemService::AuthInterface]
    ///
    /// This action is automatically called by `accounts` when an account is configured to use this auth service.
    ///
    /// Verifies that a particular user is allowed to use a particular auth service.
    ///
    /// This action allows any user who has already set an owning account with `AuthDelegate::setOwner`.
    #[action]
    fn canAuthUserSys(user: AccountNumber) {
        unimplemented!()
    }

    /// Get the accounts this auth service delegates authority to for a sender.
    #[action]
    fn getDlgsSys(sender: AccountNumber) -> Vec<AccountNumber> {
        unimplemented!()
    }

    /// Check whether a specified set of authorizer accounts are sufficient to authorize sending a
    /// transaction from a specified sender.
    ///
    /// * `sender`: The sender account for the transaction potentially being authorized.
    /// * `authorizers`: The set of accounts that have already authorized the execution of the transaction.
    ///
    /// Returns:
    /// * `true`: If the sender's owner is among the authorizers
    /// * `false`: Otherwise
    #[action]
    fn isAuthSys(
        sender: AccountNumber,
        authorizers: Vec<AccountNumber>,
        method: Option<ServiceMethod>,
    ) -> bool {
        unimplemented!()
    }

    /// Check whether a specified set of rejecter accounts are sufficient to reject (cancel) a
    /// transaction from a specified sender.
    ///
    /// * `sender`: The sender account for the transaction potentially being rejected.
    /// * `rejecters`: The set of accounts that have already authorized the rejection of the
    ///               transaction.
    ///
    /// Returns:
    /// * `true`: If the sender's owner is among the rejecters
    /// * `false`: Otherwise
    #[action]
    fn isRejectSys(
        sender: AccountNumber,
        rejecters: Vec<AccountNumber>,
        method: Option<ServiceMethod>,
    ) -> bool {
        unimplemented!()
    }

    /// Set the owner of the sender account
    ///
    /// Whenever a sender using this auth service submits a transaction, authorization
    /// for the owned account will check authorization for the owner instead.
    #[action]
    fn setOwner(owner: AccountNumber) {
        unimplemented!()
    }

    /// Create a new account with the specified name, owned by the specified `owner` account.
    ///
    /// Existing accounts will not be modified. If the `requireMatch` flag
    /// is set, then the action will fail if the account exists but is not
    /// already owned by the specified owner.
    #[action]
    fn newAccount(name: AccountNumber, owner: AccountNumber, require_match: bool) -> bool {
        unimplemented!()
    }
}

#[test]
fn verify_schema() {
    crate::assert_schema_matches_package::<Wrapper>();
}

File: packages/system/AuthDelegate/src/AuthDelegate.cpp
#include <services/system/AuthDelegate.hpp>

#include <psibase/dispatch.hpp>
#include <services/system/Accounts.hpp>
#include <services/system/AuthAny.hpp>
#include <services/system/StagedTx.hpp>
#include "services/system/Transact.hpp"

using namespace psibase;

namespace SystemService
{
   bool AuthDelegate::checkAuthSys(std::uint32_t              flags,
                                   AccountNumber              requester,
                                   AccountNumber              sender,
                                   ServiceMethod              action,
                                   std::vector<ServiceMethod> allowedActions,
                                   std::vector<Claim>         claims)
   {
      auto owner = getOwner(sender);
      check(owner.has_value(), "account does not have an owning account");

      if (requester == *owner)
         flags = (flags & ~AuthInterface::requestMask) | AuthInterface::runAsRequesterReq;

      auto _ = recurse();
      return authServiceOf(*owner).checkAuthSys(flags, requester, *owner, std::move(action),
                                                std::move(allowedActions), std::move(claims));
   }

   void AuthDelegate::canAuthUserSys(psibase::AccountNumber user)
   {
      check(getOwner(user).has_value(), "account does not have an owning account");
   }

   std::vector<AccountNumber> AuthDelegate::getDlgsSys(psibase::AccountNumber sender)
   {
      auto owner = getOwner(sender);
      check(owner.has_value(), "account does not have an owning account");
      return {*owner};
   }

   bool AuthDelegate::isAuthSys(psibase::AccountNumber              sender,
                                std::vector<psibase::AccountNumber> authorizers)
   {
      auto owner = getOwner(sender);
      check(owner.has_value(), "account does not have an owning account");
      return std::ranges::contains(authorizers, *owner);
   }

   bool AuthDelegate::isRejectSys(psibase::AccountNumber              sender,
                                  std::vector<psibase::AccountNumber> rejecters)
   {
      auto owner = getOwner(sender);
      check(owner.has_value(), "account does not have an owning account");
      return std::ranges::contains(rejecters, *owner);
   }

   void AuthDelegate::setOwner(psibase::AccountNumber owner)
   {
      auto table  = open<AuthDelegateTable>();
      auto record = table.getIndex<0>().get(owner);
      auto sender = getSender();
      if (sender == owner || record && record->owner == sender)
      {
         abortMessage("circular ownership");
      }

      to<Accounts>().incAuthSeq(sender);

      table.put(AuthDelegateRecord{.account = getSender(), .owner = owner});
   }

   std::optional<AccountNumber> AuthDelegate::getOwner(psibase::AccountNumber account)
   {
      if (auto row = open<AuthDelegateTable>(KvMode::read).getIndex<0>().get(account))
         return row->owner;
      return std::nullopt;
   }

   Actor<AuthInterface> AuthDelegate::authServiceOf(psibase::AccountNumber account)
   {
      return Actor<AuthInterface>{service, to<Accounts>().getAuthOf(account)};
   }

   // cases to consider:
   // - the account does not exist
   // - the account exists and uses AuthDelegate with another owner
   // - the account exists and uses a different auth service
   // - the account exists and uses a different auth service but there is a record in auth delegate
   bool AuthDelegate::newAccount(psibase::AccountNumber name,
                                 psibase::AccountNumber owner,
                                 bool                   requireMatch)
   {
      auto table = open<AuthDelegateTable>();
      if (requireMatch)
      {
         auto record = table.getIndex<0>().get(name);
         if (record && record->owner != owner)
         {
            abortMessage(std::format("{} is owned by {}", name.str(), record->owner.str()));
         }
      }

      check(to<Accounts>().exists(owner), "owner account does not exist");

      bool created = to<Accounts>().newAccount(name, AuthDelegate::service, requireMatch);
      if (created)
         table.put(AuthDelegateRecord{.account = name, .owner = owner});
      return created;
   }

}  // namespace SystemService

PSIBASE_DISPATCH(SystemService::AuthDelegate)
